doc/tstool.8
.Dd March 20, 2025
.Os
.Dt tstool 8
.Sh NAME
.Nm tstool
.Nd Timestamping control utility
.Sh SYNOPSIS
.Nm
.Op Fl -help
.Nm
.Op Fl -verbose | v
.Op Fl -config-file Pa file | Fl f Pa file
.Op Fl -persistent
.Op Fl -initial
.Op Fl -enable-efx
.Op Fl -disable-efx
.Ar command ...
.Sh DESCRIPTION
Examine and adjust PTP Hardware Clock devices, system clock and interface timestamping settings.
.Pp
This utility uses the mechanisms available to sfptpd to manipulate precision clocks on the system.
.Pp
Clocks may be specified by phc number, interface name or 'system'.
.Pp
A configuration file in the same format as used by
.Xr sfptpd 8
may be supplied.
Any settings in the configuration file will take precedence over the command line options.
Note that this behaviour differs from
.Nm sfptpd
which performs a second pass of the command line arguments after interpreting the config file.
The config file is useful for expressing options like
.Ql phc_diff_methods
for defining the preferred clock comparison methods to use.
.Pp
Note that trace is output from internal operations as if they were run by
.Nm sfptpd .
To suppress this,
specify
.Ql trace_level all 0
in a configuration file.
.Ss Commands
.Bl -tag interface
.It Nm clock list
List clocks.
.It Nm clock info Ar clock
Show information about
.Ar clock .
.It Nm clock get Ar clock
Read the time from
.Ar clock .
.It Nm clock step Ar clock Ar offset
Step
.Ar clock
by
.Ar offset
seconds.
.It Nm clock slew Ar clock Ar ppb
Slew
.Ar clock
by
.Ar ppb
parts per billion.
.It Nm clock set_to Ar clock1 Ar clock2
Set
.Ar clock1
to
.Ar clock2 .
.It Nm clock diff Ar clock1 Ar clock2
Report offset in seconds by which
.Ar clock1
is in advance of
.Ar clock2 .
.It Nm clock set_sync Ar clock Op Ar timeout
Set the sync status flag of supported clocks to indicate that
.Ar clock
is in sync,
expiring this flag after
.Ar timeout
seconds if not renewed.
Specify a
.Ar timeout
of
.Li 0
to clear the sync flag.
The default timeout of
.Li 60
seconds is used if the
.Ar timeout
argument is omitted.
.It Nm clock dedup
Deduplicate clock devices by working out which ones share the same underlying
physical clock.
.It Nm interface list
List available physical interfaces suitable for PTP.
.It Nm interface info Ar interface
Show information about a physical interface.
.It Nm interface set_ts Ar interface Ar tx-type Ar rx-filter
Set timestamping modes for an interface.
Use
.Nm interface info
to see what values are available to set for a given interface.
.El
.Sh BUGS
Please raise bug reports at:
.Aq https://github.com/Xilinx-CNS/sfptpd/issues
.Pp
Support for users of AMD Solarflare NICs is available from:
.Aq support-nic@amd.com
.Sh AUTHORS
Advanced Micro Devices, Inc.
.Sh SEE ALSO
.Xr sfptpd 8 .
.Pp
Full documentation is available at:
.Aq https://docs.amd.com/r/en-US/ug1602-ptp-user
.Sh COPYRIGHT
Copyright (c) 2024-2025 Advanced Micro Devices, Inc.


CHANGELOG
Changelog
---------

### v3.7.1.1000 Release Candidate

Features:
 ~ Minor command line enhancements to support container usage (SWPTP-1401)
     - Allow config to be read from stdin with '-f -'
     - Add '--console' option to redirect logs to console

Issue SWPTP-1395
 ~ Fix missing sfptpmon script in .tar.gz release package
 ~ Omit unnecessary new reference to system clock frequency correction state
   file in .rpm packages to avoid confusing 'alien' repackager tool
Issue SWPTP-1403
 ~ Ignore irrelevant PTP packets in sfptpmon event monitoring script instead of
   exiting
Issue SWPTP-1405
 ~ Avoid theoretical risk of double clock correction on startup
 ~ Extend period NIC considered to have unset time to 5 years after epoch
Issue SWPTP-1408
 ~ Fix possible risk of stack corruption in teaming netlink handler


### v3.7.0.1006 Feature Release

Features:
 ~ Allow running as non-root user with minimum required capabilities (SWPTP-442)
     - enable with 'user' configuration or --user command line options
     - root user required for chronyd clock control script to restart chronyd
     - example udev rules support hotplug with non-root user
 ~ Support multiple PPS sources (SWPTP-776)
 ~ Support bridges including over bonds in similar way to LACP bonds (SWPTP-826)
 ~ Support MACVLANs when timestamping enabled in host netns (SWPTP-992)
 ~ Allow a bond, bridge or vlan to be specified for freerun clock (SWPTP-1169)
 ~ Replace bond, team and interface probing with netlink scanning (SWPTP-1269)
 ~ Mix 'efx' clock diff method with list of preferred phc methods (SWPTP-1285)
 ~ Add sync-instances file showing factors influencing selection (SWPTP-1286)
 ~ Add ntp and servo logging modules for targetted diagnostics (SWPTP-1291)
 ~ Reflect external constraints (ntpd, chronyd) as selection rule (SWPTP-1295)
 ~ Better quality logging and trace on hotplug events (SWPTP-1300)
 ~ IP address of parent clock added to state and nodes files (SWPTP-1312)
 ~ Better init system integration with new command line options (SWPTP-1321)
     --daemon       supersedes now-deprecated 'daemon' config file option
     --no-daemon    overrides now-deprecated 'daemon' config file option
     --version      outputs version number
     --test-config  tests configuration file
     --user USER[:GROUP]
     and synchronises with systemd on startup as 'notify'-type service
 ~ Support up to 6 destinations for PTP event monitoring messages (SWPTP-1340)
 ~ Add 'sfptpmon' script to receive PTP event monitoring messages (SWPTP-1342)
 ~ Accept any port for unicast PTP event monitoring destinations (SWPTP-1346)
 ~ The 'ntpsec' implementation of ntpd available in the EPEL repositories for
     RHEL 9 is confirmed to be fully interoperable with sfptpd, allowing
     system clock disciplining to be controlled at runtime.

Removed support:
 ~ Special provision for SFN5000 and SFN6000 series adapters removed (SWPTP-1306)
 ~ Provision for hardware timestamping without Linux PHC removed (SWPTP-1306)
 ~ Support, testing and packaging for RHEL6 is dropped (SWPTP-1371)

Issue SWPTP-1278
 ~ Fix retrieval of transmit timestamps to recover from nic reset quicker
Issue SWPTP-1296
 ~ Improve logging of timer fd activity to identify system configuration issues
Issue SWPTP-1314
 ~ Fix issues using PPS via /dev/ptp external timestamping pin
Issue SWPTP-1326
 ~ Alter buffering for RT JSON stats output to avoid excessive writes per sec
Issue SWPTP-1347
 ~ Remove bogus reserved field from rx timing event monitoring TLV
   ** NOTE ** Faulty senders and receivers need updating together.
              This only applies to users of the mon_rx_sync_timing_data option.
Issue SWPTP-1358
 ~ Fix chronyd client state handling avoiding crashes after system startup
Issue SWPTP-1366
 ~ Fix ntpd client handling to avoid quitting when peer stats not available


### v3.6.0.1015 Feature Release

Features:
 ~ Support Xilinx Alveo X3522 Ethernet adapters (SWPTP-1162)

 ~ Support for operating alongside the chronyd service is now provided in a
    separate 'crny' sync module from the 'ntp' module used for ntpd, e.g.
     [general]
     sync_module crny crny1

    To aid the transition from existing behaviour under the 'ntp' sync module
    a suitable instance is created implicitly. To disable this for simplified
    logging and diagnostics, specify a sync module with no instances:
     [general]
     sync_module crny

 ~ Allow user-defined scripts to reconfigure and restart a chronyd service
    as a workaround to lack of a 'clock control' facility in chronyd. This
    brings parity in the supportable use cases of ntpd and chronyd. (SWPTP-1216)
     [crny]
     control_script /usr/share/doc/sfptpd/examples/chrony_clockcontrol.py

   ** N.B. restarting chronyd is destabilising for NTP synchronisation so this
    feature should be avoided when chrony's time sync performance influences
    sync instance selection. In the absence of native runtime control for
    chronyd it is best used in an always-on or always-off mode with sfptpd. If
    chronyd control is required for fallback it is advisable to give it a
    lower (numerically higher) 'priority' setting so it is only selected on
    total failure of PTP and to increase 'selection_holdoff_period' to 60s.
    Similarly, it would not be recommended to use this clock control method
    while chrony was operating as a clustering determinant (i.e. discriminator).

 ~ Allow operation but block system clock updates when an uncontrollable
    chronyd service is running, except when this condition holds at startup.
    This is improved behaviour when the conflict is transient. It is, however,
    recommended in such situations to configure sfptpd explicitly to treat
    the system clock as read only. To disable this check at startup,
    configure as follows. (SWPTP-1266)
     [general]
     ignore_critical clock-control-conflict

 ~ Allow operation in the absence of a kernel PTP subsystem or PTP and PPS

src/tstool/tstool.c
/* SPDX-License-Identifier: BSD-3-Clause */
/* (c) Copyright 2024 Advanced Micro Devices, Inc. */

/* Timestamping control utility */

#include <assert.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <regex.h>
#include <unistd.h>
#include <getopt.h>
#include <errno.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <fcntl.h>
#include <linux/sockios.h>
#include <linux/net_tstamp.h>

#include "sfptpd_config.h"
#include "sfptpd_general_config.h"
#include "sfptpd_netlink.h"
#include "sfptpd_clock.h"
#include "sfptpd_interface.h"
#include "sfptpd_logging.h"


/****************************************************************************
 * Constant
 ****************************************************************************/

#define OPT_PERSISTENT  0x10000
#define OPT_INITIAL     0x10001
#define OPT_ENABLE_EFX  0x10002
#define OPT_DISABLE_EFX 0x10003

static const char *opts_short = "+hvf:";
static const struct option opts_long[] = {
	{ "help", 0, NULL, (int) 'h' },
	{ "verbose", 0, NULL, (int) 'v' },
	{ "config-file", 1, NULL, (int) 'f' },
	{ "persistent", 0, NULL, OPT_PERSISTENT },
	{ "initial", 0, NULL, OPT_INITIAL },
	{ "enable-efx", 0, NULL, OPT_ENABLE_EFX },
	{ "disable-efx", 0, NULL, OPT_DISABLE_EFX },
	{ NULL, 0, NULL, 0 }
};

const char *tx_types[] = {
	"off",
	"on",
	"onestep-sync",
	"onestep-p2p",
};
#define NUM_TX_TYPES (sizeof tx_types / sizeof *tx_types)

const char *rx_filters[] = {
	"none",
	"all",
	"some",
	"ptp-v1-l4-event",
	"ptp-v1-l4-sync",
	"ptp-v1-l4-delay-req",
	"ptp-v2-l4-event",
	"ptp-v2-l4-sync",
	"ptp-v2-l4-delay-req",
	"ptp-v2-l2-event",
	"ptp-v2-l2-sync",
	"ptp-v2-l2-delay-req",
	"ptp-v2-event",
	"ptp-v2-sync",
	"ptp-v2-delay-req",
	"ptp-ntp-all",
};
#define NUM_RX_FILTERS (sizeof rx_filters / sizeof *rx_filters)

const char *sof[] = {
	"tx_hardware",
	"tx_software",
	"rx_hardware",
	"rx_software",
	"software",
	"sys_hardware",
	"raw_hardware",
	"opt_id",
	"tx_sched",
	"tx_ack",
	"opt_cmsg",
	"opt_tsonly",
	"opt_stats",
	"opt_pktinfo",
	"opt_tx_swhw",
	"bind_phc",
	"opt_id_tcp",
	"17",
	"18",
	"19",
	"20",
	"21",
	"22",
	"onload_stream",
};
#define NUM_SOF (sizeof sof / sizeof *sof)


/****************************************************************************
 * Types
 ****************************************************************************/

enum clock_command_e {
	CLOCK_CMD_LIST,
	CLOCK_CMD_INFO,
	CLOCK_CMD_GET,
	CLOCK_CMD_STEP,
	CLOCK_CMD_SLEW,
	CLOCK_CMD_SET_TO,
	CLOCK_CMD_DIFF,
	CLOCK_CMD_DEDUP,
	CLOCK_CMD_SET_SYNC,
	CLOCK_CMD_INVALID
};

struct clock_command {
	enum clock_command_e tag;
	const char *name;
	int clock_args;
};

enum intf_command_e {
	INTF_CMD_LIST,
	INTF_CMD_INFO,
	INTF_CMD_SET_TS,
	INTF_CMD_INVALID
};

struct intf_command {
	enum intf_command_e tag;
	const char *name;
	int intf_args;
};

/****************************************************************************
 * Local Data
 ****************************************************************************/

static struct sfptpd_config *config = NULL;
static struct sfptpd_nl_state *netlink = NULL;
static const struct sfptpd_link_table *initial_link_table = NULL;

static const struct clock_command clock_cmds[] = {
	{ CLOCK_CMD_LIST,     "list",     0 },
	{ CLOCK_CMD_INFO,     "info",     1 },
	{ CLOCK_CMD_GET,      "get",      1 },
	{ CLOCK_CMD_STEP,     "step",     1 },
	{ CLOCK_CMD_SLEW,     "slew",     1 },
	{ CLOCK_CMD_SET_TO,   "set_to",   2 },
	{ CLOCK_CMD_DIFF,     "diff",     2 },
	{ CLOCK_CMD_DEDUP,    "dedup",    0 },
	{ CLOCK_CMD_SET_SYNC, "set_sync", 1 },
	{ CLOCK_CMD_INVALID,  "INVALID",  0 },
};

static const struct intf_command intf_cmds[] = {
	{ INTF_CMD_LIST,    "list",    0 },
	{ INTF_CMD_INFO,    "info",    1 },
	{ INTF_CMD_SET_TS,  "set_ts",  1 },
	{ INTF_CMD_INVALID, "INVALID", 0 },
};


/****************************************************************************
 * Local functions
 ****************************************************************************/

static void format_flags(char *buf, ssize_t space,
			 const char **names, size_t num_known,
			 unsigned long flags)
{
	int i;
	int len = 0;
	const char ellipsis[] = u8"\u2026";
	ssize_t rewind = 1 + strlen(ellipsis);

	assert(space >= rewind);
	*buf = '\0';

	for (i = 0; len >= 0 && flags != 0 && space >= rewind; i++, flags >>= 1) {
		if (flags & 1) {
			if (i < num_known)
				len = snprintf(buf, space, " %s", names[i]);
			else
				len = snprintf(buf, space, " %d", i);
			if (len >= space)
				break;
			else
				space -= len, buf += len;
		}
	}

	if (flags != 0)
		snprintf(buf + space - rewind, rewind, "%s", ellipsis);
}

static int decode_option(const char **names, size_t num_known,
			 unsigned long *option, const char *text)
{
	int i;

	for (i = 0; i < num_known && strcmp(text, names[i]); i++);
	if (i == num_known) {
		ERROR("option %s invalid\n", text);
		return 1;
	}

	*option = i;
	return 0;
}


static int do_init(void)
{
	int rc;

	/* Start netlink service */
	netlink = sfptpd_netlink_init();
	if (netlink == NULL) {
		CRITICAL("could not start netlink\n");
		return EXIT_FAILURE;
	}

	rc = sfptpd_netlink_scan(netlink);
	if (rc != 0) {
		CRITICAL("scanning with netlink, %s\n",
			 strerror(rc));
		return rc;
	}

	/* Wait 5 seconds for initial link table */
	initial_link_table = sfptpd_netlink_table_wait(netlink, 1, 5000);
	if (initial_link_table == NULL) {
		CRITICAL("could not get initial link table, %s\n",
			 strerror(errno));
		return EXIT_FAILURE;
	}

	/* Start clock management */
	rc = sfptpd_clock_initialise(config, NULL);
	if (rc != 0)
		return EXIT_FAILURE;

	/* Start interface management */
	rc = sfptpd_interface_initialise(config, NULL,
					 initial_link_table);
	if (rc != 0)
		return EXIT_FAILURE;

	return 0;
}

static void do_finit(void)
{
	sfptpd_clock_shutdown();
	sfptpd_interface_shutdown(config);
	if (netlink != NULL)
		sfptpd_netlink_finish(netlink);
}

static void usage(FILE *stream)
{
	fprintf(stream,
		"syntax: %s [OPTIONS] SUBSYSTEM COMMAND..\n"
		"\n"
		"  OPTIONS\n"
		"        --persistent            Use sfptpd persistent frequency adjustment\n"
		"        --initial               Perform sfptpd initial clock correction\n"
		"    -f, --config-file=FILE      Configure from FILE, or stdin if '-'\n"
		"    -h, --help                  Show usage\n"
		"    -v, --verbose               Be verbose\n"
		"        --enable-efx            Enable SIOCEFX sfc private ioctl calls (the default)\n"
		"        --disable-efx           Disable SIOCEFX sfc private ioctl calls\n"
		"\n"
		"  CLOCK SUBSYSTEM\n"
		"    clock list                  List clocks\n"
		"    clock info CLOCK            Show clock information\n"
		"    clock get CLOCK             Read clock\n"
		"    clock step CLOCK OFFSET     Step clock\n"
		"    clock slew CLOCK PPB        Adjust clock frequency\n"
		"    clock set_to CLOCK1 CLOCK2  CLOCK1 := CLOCK2\n"
		"    clock diff CLOCK1 CLOCK2    CLOCK1 - CLOCK2\n"
		"    clock set_sync CLOCK TIMEO  Set CLOCK sync flag for TIMEO seconds\n"
		"    clock dedup                 Deduplicate shared phc devices\n\n"
		"      CLOCK := <phcN> | <ethN> | system\n\n"
		"  INTERFACE SUBSYSTEM\n"
		"    interface list              List physical interfaces\n"
		"    interface info INTF         Show interface information\n"
		"    interface set_ts TX RX      Set timestamp modes\n\n"
		"      INTF := <ethN>\n"
                "      See 'info' response for available TX and RX modes\n",
		program_invocation_short_name);
}

static int clock_command(int argc, char *argv[])
{
#define MAX_CLOCKS 2
	struct sfptpd_timespec times[MAX_CLOCKS];
	struct sfptpd_clock *clocks[MAX_CLOCKS];
	struct sfptpd_interface *interface;
	struct sfptpd_clock **all_clocks;
	const struct clock_command *cmd;
	const char *command;
	size_t num_clocks;
	sfptpd_time_t freq_adj;
	sfptpd_time_t tick_len;
	sfptpd_time_t f = 0.0L;
	int tokens;
	int rc = 0;
	int i;

	if (argc < 1) {
		usage(stderr);
		return EXIT_FAILURE;
	}

	command = argv[0];

	for (cmd = clock_cmds;
	     cmd->tag != CLOCK_CMD_INVALID &&
	     strcmp(command, cmd->name);
	     cmd++);

	assert(cmd->clock_args <= MAX_CLOCKS);
	if (argc - 1 < cmd->clock_args) {
		ERROR("insufficient number of clocks specified\n");
		usage(stderr);
		return EXIT_FAILURE;
	}

	for (i = 0; i < cmd->clock_args; i++) {
		const char *clock_ref = argv[1 + i];
		clocks[i] = sfptpd_clock_find_by_name(clock_ref);
		if (clocks[i] == NULL &&
		    (interface = sfptpd_interface_find_by_name(clock_ref))) {
			clocks[i] = sfptpd_interface_get_clock(interface);
		}
		if (clocks[i] == NULL) {
			fprintf(stderr, "unknown clock: %s\n", clock_ref);
			return EXIT_FAILURE;
		}
	}

	switch(cmd->tag) {
	case CLOCK_CMD_LIST:
		all_clocks = sfptpd_clock_get_active_snapshot(&num_clocks);
		for (i = 0; i < num_clocks; i++) {
			printf("%s\n", sfptpd_clock_get_long_name(all_clocks[i]));
		}
		sfptpd_clock_free_active_snapshot(all_clocks);
		break;
	case CLOCK_CMD_INFO:
		sfptpd_clock_get_frequency(clocks[0], &freq_adj, &tick_len);
		printf("short-name: %s\n"
		       "long-name: %s\n"
		       "hw-id: %s\n"
		       "persistent-freq-correction: %.3Lf ppb\n"
		       "max-freq-adj: %.3Lf ppb\n"
		       "freq-adj: %.3Lf ppb\n"
		       "tick-len: %.3Lf s\n"
		       "diff-method: %s\n",
		       sfptpd_clock_get_short_name(clocks[0]),
		       sfptpd_clock_get_long_name(clocks[0]),
		       sfptpd_clock_get_hw_id_string(clocks[0]),
		       sfptpd_clock_get_freq_correction(clocks[0]),
		       sfptpd_clock_get_max_frequency_adjustment(clocks[0]),
		       freq_adj,
		       tick_len / 1000000000.0L,
		       sfptpd_clock_get_diff_method(clocks[0]));
		break;
	case CLOCK_CMD_GET:
		rc = sfptpd_clock_get_time(clocks[0], times + 0);
		printf("%s: " SFPTPD_FMT_SFTIMESPEC "\n",
		       sfptpd_clock_get_short_name(clocks[0]),
		       SFPTPD_ARGS_SFTIMESPEC(times[0]));
		break;
	case CLOCK_CMD_STEP:
		tokens = sscanf(argv[1 + cmd->clock_args], "%Lf", &f);
		if (tokens != 1) {
			ERROR("invalid offset specified\n");
			return EXIT_FAILURE;
		}
		sfptpd_time_float_s_to_timespec(f, times + 0);
		rc = sfptpd_clock_adjust_time(clocks[0], times + 0);
		break;
	case CLOCK_CMD_SLEW:
		tokens = sscanf(argv[1 + cmd->clock_args], "%Lf", &f);
		if (tokens != 1) {
			ERROR("invalid frequency adjustment specified\n");
			return EXIT_FAILURE;
		}
		rc = sfptpd_clock_adjust_frequency(clocks[0], f);
		break;
	case CLOCK_CMD_DIFF:
		rc = sfptpd_clock_compare(clocks[0], clocks[1], times + 0);
		printf("%s-%s: " SFPTPD_FMT_SSFTIMESPEC "\n",
		       sfptpd_clock_get_short_name(clocks[0]),
		       sfptpd_clock_get_short_name(clocks[1]),
		       SFPTPD_ARGS_SSFTIMESPEC(times[0]));
		break;
	case CLOCK_CMD_SET_TO:
		rc = sfptpd_clock_set_time(clocks[0], clocks[1], NULL, false);
		break;
	case CLOCK_CMD_SET_SYNC:
		if (argc > 1 + cmd->clock_args)
			tokens = sscanf(argv[1 + cmd->clock_args], "%Lf", &f);
		else
			f = SFPTPD_STATS_CONVERGENCE_MIN_PERIOD_DEFAULT;
		rc = sfptpd_clock_set_sync_status(clocks[0],
						  f ? 1 : 0,
						  f ? f : 0);
		break;
	case CLOCK_CMD_DEDUP:
		rc = sfptpd_clock_deduplicate();
		break;
	default:
		fprintf(stderr, "unknown clock command: %s\n", command);
		usage(stderr);
		return EXIT_FAILURE;
	}

	if (rc != 0) {
		ERROR("tstool: clock: %s: %s\n", command, strerror(rc));
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}


static int intf_command(int argc, char *argv[])
{
#define MAX_INTFS 2
	struct sfptpd_interface *interfaces[MAX_INTFS];
	struct sfptpd_db_query_result query_result;
	struct hwtstamp_config so_ts_req = { 0 };
	struct ethtool_ts_info ts_info;
	struct sfptpd_interface *intf;
	struct sfptpd_clock *clock;
	const struct intf_command *cmd;
	const char *command;
	char tx_types_str[128];
	char rx_filters_str[128];
	char sof_str[128];
	char tx_type_str[20];
	char rx_filter_str[20];
	bool supports_efx;
	bool supports_phc;
	unsigned long rx;
	unsigned long tx;
	int device_idx;
	int rc = 0;
	int i;

	if (argc < 1) {
		usage(stderr);
		return EXIT_FAILURE;
	}

	command = argv[0];

	for (cmd = intf_cmds;
	     cmd->tag != INTF_CMD_INVALID &&
	     strcmp(command, cmd->name);
	     cmd++);

	assert(cmd->intf_args <= MAX_INTFS);
	if (argc - 1 < cmd->intf_args) {
		ERROR("insufficient number of interfaces specified\n");
		usage(stderr);
		return EXIT_FAILURE;
	}

	for (i = 0; i < cmd->intf_args; i++) {
		const char *intf_ref = argv[1 + i];
		interfaces[i] = sfptpd_interface_find_by_name(intf_ref);
		if (interfaces[i] == NULL) {
			fprintf(stderr, "unknown interfface: %s\n", intf_ref);
			return EXIT_FAILURE;
		}
	}

	switch(cmd->tag) {
	case INTF_CMD_LIST:
		query_result = sfptpd_interface_get_active_ptp_snapshot();
		for (i = 0; i < query_result.num_records; i++) {
			struct sfptpd_interface **intfp = query_result.record_ptrs[i];

			intf = *intfp;
			printf("%s\n", sfptpd_interface_get_name(intf));
		}
		query_result.free(&query_result);
		break;
	case INTF_CMD_INFO:
		intf = interfaces[0];
		clock = sfptpd_interface_get_clock(intf);
		sfptpd_interface_get_clock_device_idx(intf, &supports_phc, &device_idx, &supports_efx);
		sfptpd_interface_get_ts_info(intf, &ts_info);
		sfptpd_interface_ioctl(intf, SIOCGHWTSTAMP, &so_ts_req);
		format_flags(tx_types_str, sizeof tx_types_str, tx_types, NUM_TX_TYPES, ts_info.tx_types);
		format_flags(rx_filters_str, sizeof rx_filters_str, rx_filters, NUM_RX_FILTERS, ts_info.rx_filters);
		format_flags(sof_str, sizeof sof_str, sof, NUM_SOF, ts_info.so_timestamping);
		format_flags(tx_type_str, sizeof tx_types_str, tx_types, NUM_TX_TYPES, 1 << so_ts_req.tx_type);
		format_flags(rx_filter_str, sizeof rx_filters_str, rx_filters, NUM_RX_FILTERS, 1 << so_ts_req.rx_filter);
		printf("interface: %s\n"
		       "clock: %s\n"
		       "mac-address: %s\n"
		       "fw-version: %s\n"
		       "supported-apis:%s%s\n"
		       "supported-tx-modes:%s\n"
		       "supported-rx-filters:%s\n"
		       "supported-sof-flags:%s\n"
		       "tx-mode:%s\n"
		       "rx-filter:%s\n",
		       sfptpd_interface_get_name(intf),
		       clock == NULL ? "none" : sfptpd_clock_get_short_name(clock),
		       sfptpd_interface_get_mac_string(intf),
		       sfptpd_interface_get_fw_version(intf),
		       supports_phc ? " phc" : "",
		       supports_efx ? " efx" : "",
		       tx_types_str,
		       rx_filters_str,
		       sof_str,
		       tx_type_str,
		       rx_filter_str);
		break;
	case INTF_CMD_SET_TS:
		intf = interfaces[0];
		if (argc < 3 + cmd->intf_args) {
			ERROR("insufficient arguments for timestamp configuration\n");
			return EXIT_FAILURE;
		}
		if (decode_option(tx_types, NUM_TX_TYPES, &tx, argv[1 + cmd->intf_args]))
			return EXIT_FAILURE;
		if (decode_option(rx_filters, NUM_RX_FILTERS, &rx, argv[2 + cmd->intf_args]))
			return EXIT_FAILURE;
		so_ts_req.tx_type = tx;
		so_ts_req.rx_filter = rx;
		rc = sfptpd_interface_ioctl(intf, SIOCSHWTSTAMP, &so_ts_req);
		break;
	default:
		fprintf(stderr, "unknown interface command: %s\n", command);
		usage(stderr);
		return EXIT_FAILURE;
	}

	if (rc != 0) {
		ERROR("tstool: interface: %s: %s\n", command, strerror(rc));
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}


/****************************************************************************
 * Global functions
 ****************************************************************************/

int main(int argc, char *argv[])
{
	struct sfptpd_config_general *gconf;
	const char *subsystem;
	int index;
	int opt;
	int rc = EXIT_FAILURE;
	int ret;

	/* Initialise config */
	ret = sfptpd_config_create(&config);
	if (ret != 0)
		return EXIT_FAILURE;

	/* Tweak config */
	gconf = sfptpd_general_config_get(config);
	gconf->non_sfc_nics = true;
	gconf->timestamping.disable_on_exit = false;
	gconf->clocks.persistent_correction = false;
	gconf->clocks.no_initial_correction = true;
	gconf->avoid_efx = false;

	/* Handle command line arguments */
	while ((opt = getopt_long(argc, argv, opts_short, opts_long, &index)) != -1) {
		switch (opt) {
		case 'f':
			sfptpd_config_set_config_file(config, optarg);
			break;
		case 'h':
			usage(stdout);
			goto fail;
		case 'v':
			sfptpd_log_set_trace_level(SFPTPD_COMPONENT_ID_NETLINK, 3);
			sfptpd_log_set_trace_level(SFPTPD_COMPONENT_ID_SFPTPD, 6);
			sfptpd_log_set_trace_level(SFPTPD_COMPONENT_ID_CLOCKS, 3);
			break;
		case OPT_PERSISTENT:
			gconf->clocks.persistent_correction = true;
			break;
		case OPT_INITIAL:
			gconf->clocks.no_initial_correction = false;
			break;
		case OPT_ENABLE_EFX:
			gconf->avoid_efx = false;
			break;
		case OPT_DISABLE_EFX:
			gconf->avoid_efx = true;
			break;
		default:
			fprintf(stderr, "unexpected option: %s\n", argv[optind]);
			usage(stderr);
			goto fail;
		}
	}

	if (argc - optind < 1) {
		usage(stderr);
		goto fail;
	}

	/* Parse the configuration file */
	rc = sfptpd_config_parse_file(config);
	if (rc != 0)
		goto fail;

	if (do_init() != 0)
		goto fail;

	subsystem = argv[optind++];

	if (!strcmp(subsystem, "clock")) {
		rc = clock_command(argc - optind, argv + optind);
	} else if (!strcmp(subsystem, "interface") ||
		   !strcmp(subsystem, "intf")) {
		rc = intf_command(argc - optind, argv + optind);
	} else {
		fprintf(stderr, "unknown subsystem: %s\n", subsystem);
		usage(stderr);
	}

	do_finit();

fail:
	sfptpd_config_destroy(config);

	return rc;
}
